pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -0,0 +1,172 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// TODO: put constant value instead of labels
+// TODO: watch also the case of Namespace Creation, now is not managed, important !!!
+
+type NamespaceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+const (
+	mappingLabel          = "mapping.liqo.io"
+	offloadingLabel       = "offloading.liqo.io"
+	offloadingPrefixLabel = "offloading.liqo.io/"
+)
+
+// TO EVALUATE: when is necessary to update namespace natting table (NNT), if the name exposed in mapping.liqo.io label,
+// is different from the already present, client will change the label value to the old one. For users is not possible
+// to change names of remote namespaces that have already been created with a certain name.
+
+func (r *NamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	// TODO: add namespace Finalizer
+	var namespace v1.Namespace
+	if err := r.Get(ctx, req.NamespacedName, &namespace); err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("---------------------- I have to delete all entries from NNT, if present")
+			return ctrl.Result{}, nil
+		} else {
+			klog.Error(err, " unable to fetch Namespace for some reasons")
+			return ctrl.Result{}, err
+		}
+	}
+
+	labels := namespace.Labels
+
+	// TODO : var "erase" will become a map/vector with some references to all virtual nodes on which, remote namespace
+	//        mustn't be present
+	erase := true
+
+	// 1. If mapping.liqo.io label is not present there are no remote namespaces associated with this one, erase is full
+	if _, ok := labels[mappingLabel]; ok {
+
+		// 2.a If offloading.liqo.io is present there are remote namespaces on all virual nodes
+		if _, ok := labels[offloadingLabel]; ok {
+			erase = false // TODO : erase must be empty, i will have remote namespaces on all virtual nodes
+			klog.Infof("---------------------- I have to create remote namespaces on all virtual nodes, if they aren't already present")
+		} else {
+
+			klog.Infof("---------------------- Watch for virtual-nodes labels")
+
+			// 2.b Iterate on all nodes labels (next step only virtual nodes), if the namespace has all the requested
+			// labels, is necessary to create the remote namespace on the remote cluster associated with the virtual
+			// node
+
+			nodes := &v1.NodeList{}
+			if err := r.List(context.Background(), nodes, client.MatchingLabels{"type": "virtual-node"}); err != nil {
+				//if err := r.List(context.Background(), nodes); err != nil {
+				klog.Error(err, "Unable to list virtual nodes")
+				return ctrl.Result{}, err
+			}
+
+			for _, node := range nodes.Items {
+				i := 0
+				found := false
+				dim := len(node.Labels)
+				for key := range node.Labels {
+					i++
+					// I have to check only the offloading.liqo.io/ labels
+					// virtual nodes will always have these offloading.liqo.io/ labels
+					if len(key) > 18 && key[0:19] == offloadingPrefixLabel {
```

Comment:
You can use the `strings.StartWith` instead of this hardcoded check

Reply:
Thank you, i have changed it.